_posts/2020-12-11-torch-0.2.0-released/torch-0.2.0-released.Rmd

```diff
@@ -0,0 +1,310 @@
+
+- Initial support for JIT tracing
+- Multiworker dataloaders
+- Print methods for `nn_modules`
+
+## Multi-worker dataloaders
+
+`dataloaders` now respond to the `num_workers` argument and
+will run the pre-processing in parallel workers. 
+
+For example if we have the following dummy dataset that does
+a long computation:
+
+```{r}
+library(torch)
+dat <- dataset(
+  "mydataset",
+  initialize = function(time, len = 10) {
+    self$time <- time
+    self$len <- len
+  },
+  .getitem = function(i) {
+    Sys.sleep(self$time)
+    torch_randn(1)
+  },
+  .length = function() {
+    self$len
+  }
+)
+ds <- dat(1)
+system.time(ds[1])
+```
+```
+   user  system elapsed 
+  0.029   0.005   1.027 
+```
+
+We will now create 2 dataloaders, one that executes
+sequentially and another executing in parallel.
+
+```{r}
+seq_dl <- dataloader(ds, batch_size = 5)
+par_dl <- dataloader(ds, batch_size = 5, num_workers = 2)
+```
+
+We can now compare the time to take 2 batches sequentially
+and the time that it takes in parallel:
+
+```{r}
+seq_it <- dataloader_make_iter(seq_dl)
+par_it <- dataloader_make_iter(par_dl)
+
+two_batches <- function(it) {
+  dataloader_next(it)
+  dataloader_next(it)
+  "ok"
+}
+
+system.time(two_batches(seq_it))
+system.time(two_batches(par_it))
+```
+```
+   user  system elapsed 
+  0.098   0.032  10.086 
+   user  system elapsed 
+  0.065   0.008   5.134 
+```
+
+Note that batches are obtained in parallel, not individual observations. This is to support, in the future, other kinds
+of datasets that might have variable batch sizes.
+
+Using multiple workers is **not** necessarily faster than serial execution because there's a considerable overhead 
+when passing tensors from a worker to the main session as
+well as when initializing the workers. 
+
+This feature is powered by the powerful [{callr}](https://github.com/r-lib/callr) package 
+and works in all OS's supported by {torch}. [{callr}](https://github.com/r-lib/callr) let's
+us create persistent R sessions thus, we only pay once the overhead of transferring potentially large datasets
+objects to workers.
+
+In the process of implementing this feature we have made 
+dataloaders behave like [{coro}](https://github.com/r-lib/coro) iterators. 
+This means that you can now use [{coro}](https://github.com/r-lib/coro)'s syntax 
+to loop trough the dataloaders:
+
+```{r}
+coro::loop(for(batch in par_dl) {
+  print(batch$shape)
+})
+```
+
+```
+[1] 5 1
+[1] 5 1
+```
+
+This is the first {torch} release including the multiworker
+dataloaders feature and you might run into edge cases when
+using it. Do let us know if you find any problem.
+
+## Initial JIT support
+
+Programs that make use of the {torch} package are inevitably
+R programs thus, they always need an R installation in order
+to execute.
+
+As of version 0.2.0, {torch} allows users to JIT *trace* 
+{torch} R functions into TorchScript. The nice thing about
+this is that TorchScript programs are easily serializable, 
```

Comment:
```suggestion
this is that TorchScript programs are easily serializable and 
```

Reply:
serializable, optimizable and they... seems strange to change to serializable and optimizable and they